kanidm_unix_int/src/cache.rs

```diff
@@ -264,17 +266,36 @@ impl CacheLayer {
             .map_err(|e| {
                 error!("time conversion error - ex_time less than epoch? {:?}", e);
             })?;
+        // WIP #392: check user/default shell
+        // check if the provided `shell` exists on the system.
+        let mut new_token = token.clone();
+        let user_shell = new_token.shell.unwrap_or_default();
```

Comment:
I don't think this is very clear - UnixUserToken doesn't define a default for the `shell: Option<String>`, which means this will hit `https://doc.rust-lang.org/std/string/struct.String.html#impl-Default` so it will return an empty string. It's a weird behaviour to rely on, so it's likely better to change this logic to show the possibilities more clearly:

* The user did not request a shell, so we set our fall back
* The requested shell does not exist, so we set our fall back
* The requested shell does exist, so we respect it.

```
let requested_shell_exists: bool = new_token.shell
    .as_ref()
    .map(|shell| {
        let exists = Path::new(&shell).exists();
        if !exists {
            info!("User requested shell is not present on this system - {}", shell)
        }
        exists
    })
    unwrap_or_else(|| {
        info!("User has not specified a shell, using default");
        false
    });

if !requested_shell_exists {
    token.shell = Some(self.default_shell.clone())
}
```

Reply:
Took me some time to understand. Interesting. :-)
Because I saw `Path::new()` with None does not work, but seems like the map just returns None, that is captured by `.unwrap_or_else()`. As a non-expert, this is also not so obvious, but ok. As I really like your list of cases above, I was looking for the "Nothing requested" case, just to mention.